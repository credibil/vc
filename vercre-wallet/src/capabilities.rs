pub mod delay;
pub mod signer;
pub mod store;
