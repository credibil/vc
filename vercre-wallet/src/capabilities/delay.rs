//! Delay Capability
//!
//! This capability allows the application to delay the dispatch of an event by a
//!  specified amount of time.

use crux_core::capability::{CapabilityContext, Operation};
use crux_core::macros::Capability;
use serde::{Deserialize, Serialize};

/// Operations supported (by the Delay capability).
#[allow(clippy::module_name_repetitions)]
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct DelayOperation {
    /// Delay in milliseconds.
    pub delay_ms: u64,
}

/// Output.
impl Operation for DelayOperation {
    type Output = ();
}

/// Delay capability context.
#[derive(Capability)]
#[allow(clippy::module_name_repetitions)]
pub struct Delay<Ev> {
    context: CapabilityContext<DelayOperation, Ev>,
}

impl<Ev> Delay<Ev>
where
    Ev: 'static,
{
    /// Create a new Delay capability context.
    #[must_use]
    pub const fn new(context: CapabilityContext<DelayOperation, Ev>) -> Self {
        Self { context }
    }

    /// Dispatch the specified event after a timer delay.
    pub fn start(&self, ms: u64, event: Ev)
    where
        Ev: Send + 'static,
    {
        self.context.spawn({
            let ctx = self.context.clone();

            async move {
                ctx.request_from_shell(DelayOperation { delay_ms: ms }).await;
                ctx.update_app(event);
            }
        });
    }
}
