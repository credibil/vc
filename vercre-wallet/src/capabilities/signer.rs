//! # Signer
//!
//! The `Signer` capability is used to sign requests and responses using the wallet's
//! signing key.
//!
//! Signing operations occur during:
//!
//! 1. Credential issuance — the wallet signs a Proof (JWT or CWT) as proof of possesion
//!    of key material. That is, proof that the wallet controls the private key for the
//!    DID the issued credential will be bound to.
//!
//! 2. Credential presentation — the wallet signs a Proof eincluded in the presentation
//!    response as proof of possession of key material. That is, proof that the wallet
//!    controls the private key for the DID the presentation response is bound to.

use std::fmt::Display;

use base64ct::{Base64UrlUnpadded, Encoding};
use crux_core::capability::{CapabilityContext, Operation};
use crux_core::macros::Capability;
use serde::{Deserialize, Serialize};
use thiserror::Error;
use vercre_core::jwt::Jwt;

/// Errors that can be returned by the Signer capability.
#[derive(Clone, Debug, Deserialize, Error, PartialEq, Eq)]
pub enum Error {
    /// The request was invalid.
    #[error("invalid signing request {0}")]
    InvalidRequest(String),
}

// manually implement serde::Serialize
impl Serialize for Error {
    fn serialize<S>(&self, serializer: S) -> core::result::Result<S::Ok, S::Error>
    where
        S: serde::ser::Serializer,
    {
        serializer.serialize_str(self.to_string().as_ref())
    }
}

impl From<serde_json::Error> for Error {
    fn from(err: serde_json::Error) -> Self {
        Self::InvalidRequest(err.to_string())
    }
}

/// Result type for the `Signer` capability.
pub type Result<T> = std::result::Result<T, Error>;

/// Operations supported (by the `Signer` capability).
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
#[allow(clippy::module_name_repetitions)]
pub enum SignerRequest {
    /// Sign the provided message
    Sign(Vec<u8>),

    /// Verification information
    Verification,
}

/// `SignerResponse` represents the output expected from any implementer of the
/// Signer capability.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
#[allow(clippy::module_name_repetitions)]
pub enum SignerResponse {
    /// The signing operation returned the signature without issue.
    Signature(Vec<u8>),

    /// Signer's verification information.
    Verification {
        /// The algorithm used for signing.
        alg: String,

        /// The key identifier.
        kid: String,
    },

    /// The signing operation returned with the specified error.
    Err(String),
}

/// Operation provides a Crux wrapper for the `SignerRequest` consumed by the
/// Shell.
///
/// The Output type allows us to specify the expected response type from the
/// Shell.
impl Operation for SignerRequest {
    type Output = SignerResponse;
}

/// The Signer type used to implement the capability.
#[derive(Capability)]
pub struct Signer<Ev> {
    context: CapabilityContext<SignerRequest, Ev>,
}

impl<Ev> Signer<Ev>
where
    Ev: 'static,
{
    /// Create a new Signer capability context.
    #[must_use]
    pub const fn new(context: CapabilityContext<SignerRequest, Ev>) -> Self {
        Self { context }
    }

    /// Sign provided message using the vercre-wallet's signing key. Returns the signed output.
    ///
    /// Dispatches the 'callback' event specified in the capability request.
    ///
    /// # Panics
    ///
    /// Panics if the response is an `Err` variant.
    pub fn sign<T, F>(&self, jwt: &Jwt<T>, make_event: F)
    where
        T: Serialize + Clone,
        Jwt<T>: Display,
        F: Fn(Result<String>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let ctx = self.context.clone();
            let msg = jwt.to_string();

            async move {
                let request = SignerRequest::Sign(msg.clone().into_bytes());

                match ctx.request_from_shell(request).await {
                    SignerResponse::Signature(sig) => {
                        let sig_enc = Base64UrlUnpadded::encode_string(&sig);
                        let signed = format!("{msg}.{sig_enc}");
                        ctx.update_app(make_event(Ok(signed)));
                    }
                    SignerResponse::Err(err) => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(err))));
                    }
                    SignerResponse::Verification { .. } => {
                        panic!("Verification is not a valid response");
                    }
                }
            }
        });
    }

    /// Get the signer's verification information.
    ///
    /// Dispatches the 'callback' event specified in the capability request.
    ///
    /// # Panics
    ///
    /// Panics if the response is a `Signature` variant.
    pub fn verification<F>(&self, make_event: F)
    where
        F: Fn(Result<(String, String)>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let ctx = self.context.clone();

            async move {
                let request = SignerRequest::Verification;

                match ctx.request_from_shell(request).await {
                    SignerResponse::Verification { alg, kid } => {
                        ctx.update_app(make_event(Ok((alg, kid))));
                    }
                    SignerResponse::Err(err) => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(err))));
                    }
                    SignerResponse::Signature(_) => panic!("Signature is not a valid response"),
                }
            }
        });
    }
}
