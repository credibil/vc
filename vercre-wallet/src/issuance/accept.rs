//! # Endpoint to accept an offer
//! 
//! Used to update the issuance status to `Accepted` when the Holder has accepted an offer, or to
//! `PendingPin` when a user has accepted an offer and a PIN is required. To reject an offer and
//! clear the issuance state, use the reset endpoint.

use std::fmt::Debug;

use tracing::instrument;
use vercre_core::error::Err;
use vercre_core::provider::{Callback, Signer, StateManager};
use vercre_core::{err, Result};

use crate::issuance::{Issuance, Status};
use crate::storer::CredentialStorer;
use crate::{Endpoint, Flow};

impl<P> Endpoint<P>
where
    P: Callback + Signer + StateManager + Clone + Debug + CredentialStorer,
{
    /// Accept endpoint updates the issuance status to `Accepted` when the Holder has accepted an
    /// offer, or to `PendingPin` when a user has accepted an offer and a PIN is required.
    ///
    /// # Errors
    ///
    /// Returns an error if the request is invalid or the provider is unavailable.
    #[instrument(level = "debug", skip(self))]
    pub async fn accept(&self) -> Result<()> {
        let ctx = Context {
            _p: std::marker::PhantomData,
            issuance: Issuance::default(),
        };

        vercre_core::Endpoint::handle_request(self, &(), ctx).await
    }
}

#[derive(Debug, Default)]
struct Context<P> {
    _p: std::marker::PhantomData<P>,
    issuance: Issuance,
}

impl<P> vercre_core::Context for Context<P>
where
    P: StateManager + Debug,
{
    type Provider = P;
    type Request = ();
    type Response = ();

    async fn verify(&mut self, provider: &P, _req: &Self::Request) -> Result<&Self> {
        tracing::debug!("Context::verify");

        // Check we are processing an offer and we are at the expected point in the flow.
        let Some(stashed) = provider.get_opt(&Flow::Issuance.to_string()).await? else {
            err!(Err::InvalidRequest, "no issuance in progress");
        };
        let issuance: Issuance = serde_json::from_slice(&stashed)?;
        if issuance.status != Status::Ready {
            err!(Err::InvalidRequest, "invalid issuance status");
        }
        let Some(grants) = &issuance.offer.grants else {
            err!(Err::InvalidRequest, "no grants");
        };
        if grants.pre_authorized_code.is_none() {
            err!(Err::InvalidRequest, "no pre-authorized code");
        };
        self.issuance = issuance;

        Ok(self)
    }

    async fn process(&self, provider: &P, _req: &Self::Request) -> Result<Self::Response> {
        tracing::debug!("Context::process");

        // Update the issuance status
        let mut issuance = self.issuance.clone();
        let Some(grants) = &issuance.offer.grants else {
            err!(Err::InvalidRequest, "no grants");
        };
        let Some(pre_auth_code) = &grants.pre_authorized_code else {
            err!(Err::InvalidRequest, "no pre-authorized code");
        };
        if pre_auth_code.tx_code.is_some() {
            issuance.status = Status::PendingPin;
        } else {
            issuance.status = Status::Accepted;
        }
        provider.put_opt(&Flow::Issuance.to_string(), serde_json::to_vec(&issuance)?, None).await?;

        Ok(())
    }
}
