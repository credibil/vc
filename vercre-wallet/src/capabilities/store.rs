//! # Verifiable Credential Store

use crux_core::capability::{CapabilityContext, Operation};
use crux_core::macros::Capability;
use serde::{Deserialize, Serialize};
use thiserror::Error;
use vercre_core::w3c::vp::Constraints;

use crate::credential::Credential;

/// Errors that can be returned by the Store capability.
#[derive(Clone, Debug, Deserialize, Error, PartialEq, Eq)]
pub enum Error {
    /// The request was invalid.
    #[error("invalid store request {0}")]
    InvalidRequest(String),
}

// manually implement serde::Serialize
impl Serialize for Error {
    fn serialize<S>(&self, serializer: S) -> core::result::Result<S::Ok, S::Error>
    where
        S: serde::ser::Serializer,
    {
        serializer.serialize_str(self.to_string().as_ref())
    }
}

/// Result type for the Store capability.
pub type Result<T> = std::result::Result<T, Error>;

/// Operations supported (by the Store capability).
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
#[allow(clippy::module_name_repetitions)]
pub enum StoreRequest {
    /// Add a credential to the store
    Add(String, Vec<u8>),

    /// List credentials in the store
    List,

    /// Delete a credential from the store
    Delete(String),
}

/// `StoreResponse` represents the output expected from any implementer of the
/// Store capability.
#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
#[allow(clippy::module_name_repetitions)]
pub enum StoreResponse {
    /// The store operation returned without error and, optionally, one or more
    /// credentials.
    Ok,

    /// The store operation returned a list of credentials.
    List(Vec<u8>),

    // The store operation returned with the specified error.
    Err(String),
}

/// `Operation` provides a Crux wrapper for the `StoreRequest` consumed by the
/// Shell.
///
/// The Output type allows us to specify the expected response type from the
/// Shell.
impl Operation for StoreRequest {
    type Output = StoreResponse;
}

/// The Store type used to implement the capability.
#[derive(Capability)]
pub struct Store<Ev> {
    context: CapabilityContext<StoreRequest, Ev>,
}

impl<Ev> Store<Ev>
where
    Ev: 'static,
{
    /// Create a new Store capability context.
    #[must_use]
    pub const fn new(context: CapabilityContext<StoreRequest, Ev>) -> Self {
        Self { context }
    }

    /// Add a credential to the store.
    /// Typically the bytes would be a value serialized/deserialized by the app.
    ///
    /// Dispatches the App 'callback' event specified when the capability was
    /// requested.
    pub fn add<F>(&self, credential: Credential, make_event: F)
    where
        F: Fn(Result<()>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let ctx = self.context.clone();

            let Ok(ser) = serde_json::to_vec(&credential) else {
                return ctx.update_app(make_event(Err(Error::InvalidRequest(String::from(
                    "error serializing credential",
                )))));
            };

            async move {
                let request = StoreRequest::Add(credential.id, ser);

                match ctx.request_from_shell(request).await {
                    StoreResponse::Ok => ctx.update_app(make_event(Ok(()))),
                    StoreResponse::Err(err) => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(err))));
                    }
                    StoreResponse::List(_) => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(String::from(
                            "unexpected response from store",
                        )))));
                    }
                }
            }
        });
    }

    /// Query credentials matching the provided `JSONPath` expression.
    pub fn list<F>(&self, filter: Option<Constraints>, make_event: F)
    where
        F: Fn(Result<Vec<Credential>>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let ctx = self.context.clone();

            async move {
                let request = StoreRequest::List;

                match ctx.request_from_shell(request).await {
                    // all credentials stored by the shell
                    StoreResponse::List(bytes) => {
                        let list: Vec<Credential> = match serde_json::from_slice(&bytes) {
                            Ok(list) => list,
                            Err(e) => {
                                return ctx.update_app(make_event(Err(Error::InvalidRequest(
                                    format!("error deserializing credentials: {e}"),
                                ))));
                            }
                        };

                        // if no filter, return all credentials
                        let Some(constraints) = filter else {
                            return ctx.update_app(make_event(Ok(list)));
                        };

                        // otherwise, filter credentials
                        let mut matched: Vec<Credential> = vec![];

                        for credential in &list {
                            match constraints.satisfied(&credential.vc) {
                                Ok(true) => matched.push(credential.clone()),
                                Ok(false) => continue,
                                Err(e) => {
                                    return ctx.update_app(make_event(Err(Error::InvalidRequest(
                                        format!("error matching credentials: {e}"),
                                    ))));
                                }
                            }
                        }

                        ctx.update_app(make_event(Ok(matched)));
                    }
                    // error retrieving credentials
                    StoreResponse::Err(err) => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(err))));
                    }
                    // we should never get here
                    StoreResponse::Ok => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(String::from(
                            "unexpected response from store",
                        )))));
                    }
                }
            }
        });
    }

    /// Delete the credential specified by `id` from the store.
    pub fn delete<F>(&self, id: &str, make_event: F)
    where
        F: Fn(Result<()>) -> Ev + Send + Sync + 'static,
    {
        self.context.spawn({
            let ctx = self.context.clone();
            let id = id.to_string();

            async move {
                let request = StoreRequest::Delete(id);

                match ctx.request_from_shell(request).await {
                    StoreResponse::Ok => ctx.update_app(make_event(Ok(()))),
                    StoreResponse::Err(err) => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(err))));
                    }
                    StoreResponse::List(_) => {
                        ctx.update_app(make_event(Err(Error::InvalidRequest(String::from(
                            "unexpected response from store",
                        )))));
                    }
                }
            }
        });
    }
}
