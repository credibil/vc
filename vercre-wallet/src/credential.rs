//! # Credential App
//!
//! The Credential app module implements `crux::App` for credential management.

pub(crate) mod model;

use crux_core::macros::Effect;
#[cfg(feature = "typegen")]
use crux_core::macros::Export;
use crux_core::render::Render;
use crux_http::Http;
use serde::{Deserialize, Serialize};
#[allow(clippy::module_name_repetitions)]
pub use vercre_core::metadata::{CredentialConfiguration, CredentialDisplay};
use vercre_core::w3c::vp::Constraints;
#[allow(clippy::module_name_repetitions)]
pub use vercre_core::w3c::VerifiableCredential;

use crate::capabilities::delay::Delay;
use crate::capabilities::signer::Signer;
use crate::capabilities::store::{self, Store};
#[allow(clippy::module_name_repetitions)]
pub use crate::credential::model::{Credential, Logo};

/// App implements `crux::App` for the Credential management.
#[derive(Default)]
pub struct App;

/// Events supported by the Wallet app.
///
/// Sub-app events are delegated to the appropriate sub-app through top-level
/// wrapper events.
///
/// Events may be external (initiated by the shell), internal (initiated by
/// local processing), or both. Local events are not published
/// (`#[serde(skip)]`) to the shell.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename = "CredentialEvent")]
#[allow(clippy::large_enum_variant)]
pub enum Event {
    // -----------------------
    // Shell-initiated actions
    // -----------------------
    /// Request from shell to list stored credentials. An optional set of
    /// constraints may be provided to filter the credentials.
    List,

    // -----------------------
    // Capability callbacks
    // -----------------------
    /// An internal event to find credentials matching a Presentation Request.
    #[serde(skip)]
    ListLocal(Option<Constraints>),

    /// An internal event to set credentials fetched in `Find` into the model.
    #[serde(skip)]
    Listed(store::Result<Vec<model::Credential>>),

    // // FIXME: box to reduce the total size of the enum: Add(Box<Credential>),
    // /// Add adds a Credential to the store.
    // #[serde(skip)]
    // Add(Credential),

    // /// Added is set after a credential has been saved to the store.
    // #[serde(skip)]
    // Added(store::Result<()>),
    /// Shell-initiated event to delete a credential. The `String` is the
    /// credential `id`.
    Delete(String),

    /// An internal event initated on completion of a `Delete` event.
    #[serde(skip)]
    Deleted(store::Result<()>),

    /// Fail is set when an error occurs.
    #[serde(skip)]
    Fail(String),
}

/// Model holds an internal representation of app state.
#[derive(Default, Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
pub struct Model {
    filter: Option<Constraints>,
    credentials: Vec<model::Credential>,
    error: Option<String>,
}

/// The view model represents the App's 'external' state to the shell.
#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
#[serde(rename = "CredentialView")]
pub struct ViewModel {
    pub credentials: Vec<model::Credential>,

    #[serde(skip)]
    pub error: Option<String>,
}

/// App `Capabilities` allow the app to interface with the shell and external
/// systems.
///
/// They create `effects` (technically side effects) that are handled by the
/// shell, with the manner of handling unique to a particualr shell.
#[cfg_attr(feature = "typegen", derive(Export))]
#[derive(Effect)]
#[effect(app = "App")]
pub struct Capabilities {
    /// The Render capability allows the app to request a re-render of the UI.
    pub render: Render<Event>,

    /// The Http capability allows the app to make HTTP requests.
    pub http: Http<Event>,

    /// The Store capability allows the app to store and retrieve credentials.
    pub store: Store<Event>,

    /// The Signer capability allows the app to sign and verify messages.
    pub signer: Signer<Event>,

    /// The Delay capability allows the app to delay processing.
    pub delay: Delay<Event>,
}

impl crux_core::App for App {
    type Capabilities = Capabilities;
    type Event = Event;
    type Model = Model;
    type ViewModel = ViewModel;

    // Called in response to an event, usually by the shell but can also be in
    // response to a capability executing, or an internal processing step.
    fn update(&self, event: Self::Event, model: &mut Self::Model, caps: &Self::Capabilities) {
        match event {
            Event::List => {
                caps.store.list(None, Event::Listed);
                model.filter = None;
            }
            Event::ListLocal(filter) => {
                caps.store.list(filter.clone(), Event::Listed);
                model.filter = filter;
            }
            Event::Listed(Ok(response)) => {
                model.credentials = response;
            }
            Event::Delete(id) => {
                caps.store.delete(&id, Event::Deleted);
            }
            Event::Deleted(Ok(())) => {
                // re-query stored credentials
                self.update(Event::List, model, caps);
            }

            // ----------------------------------------------------------------
            // Error handling
            // ----------------------------------------------------------------
            Event::Listed(Err(e)) => {
                let msg = format!("Issue retrieving credentials: {e:?}");
                self.update(Event::Fail(msg), model, caps);
            }
            Event::Deleted(Err(e)) => {
                let msg = format!("Issue deleting credential: {e:?}");
                self.update(Event::Fail(msg), model, caps);
            }
            Event::Fail(msg) => {
                log::error!("{}", msg);
                model.error = Some(msg);
            }
        }

        caps.render.render();
    }

    // `view` is called by the shell to render the current state of the app.
    // Typically, this is invoked by the `render()` method of the Render
    // capability.
    fn view(&self, model: &Self::Model) -> Self::ViewModel {
        ViewModel {
            credentials: model.credentials.clone(),
            error: model.error.clone(),
        }
    }
}

#[cfg(test)]
mod tests {
    use assert_let_bind::assert_let;
    use crux_core::testing::AppTester;

    // use insta::{assert_yaml_snapshot, internals::Redaction};
    use super::*;
    use crate::capabilities::store::StoreResponse;

    /// Test that a `Event::List(filter)` event causes the app to fetch the
    /// credentials specified in the query.
    #[tokio::test]
    async fn find() {
        let app = AppTester::<App, _>::default();
        let mut model = Model::default();

        let mut update = app.update(Event::List, &mut model);

        // make real metadata request
        assert_let!(Effect::Store(request), &mut update.effects[0]);
        assert_eq!(request.operation, store::StoreRequest::List);

        let credentials = Vec::<Credential>::new();
        let values = serde_json::to_vec(&credentials).expect("should serialize");
        let response = StoreResponse::List(values.clone());

        let update = app.resolve(request, response).expect("should resolve");
        assert_eq!(update.events[0], Event::Listed(Ok(credentials.clone())));
    }
}
