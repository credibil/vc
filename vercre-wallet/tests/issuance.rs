mod issuer;

use assert_let_bind::assert_let;
use axum_test::http::header::{HeaderValue, AUTHORIZATION};
use crux_core::testing::AppTester; // assert_effect,
use crux_http::protocol::HttpResponse;
use crux_http::testing::ResponseBuilder;
use http_types::Body;
use insta::assert_yaml_snapshot as assert_snapshot;
use serde_json::{json, Value};
use test_utils::vci_provider::NORMAL_USER;
use test_utils::wallet;
use vercre_core::vci::{
    CredentialResponse, InvokeResponse, MetadataResponse, TokenRequest, TokenResponse,
};
use vercre_wallet::capabilities::signer::{SignerRequest, SignerResponse};
use vercre_wallet::capabilities::store::StoreResponse;
use vercre_wallet::credential;
use vercre_wallet::issuance::*;

/// Test that a `Offer` event causes the app to fetch credentials
/// in the offer.
#[tokio::test]
async fn receive_offer() {
    let issuer = issuer::new();

    // instantiate our app via the test harness
    let app = AppTester::<App, _>::default();
    let mut model = Model::default();

    // create InvokeRequest
    let body = json!({
        "credentials": ["EmployeeID_JWT"],
        "holder_id": NORMAL_USER,
        "pre-authorize": true,
        "user_pin_required": true,
        "callback_id": "1234"
    });

    let resp = issuer.post("/invoke").expect_success().json(&body).await;

    // generate issuer offer to 'send' to the app
    let offer_resp = resp.json::<InvokeResponse>();
    let offer = offer_resp.credential_offer.expect("should have offer");

    // ------------------------------------------------
    // Event::Offer
    //   1. saves offer to model
    //   2. creates an Issuer MetadataRequest
    //   3. emits Event::Metadata
    // ------------------------------------------------
    let offer_str = serde_json::to_string(&offer).expect("offer should serialize to string");

    let mut update = app.update(Event::Offer(offer_str), &mut model);
    assert_snapshot!("offer", app.view(&model),{
        ".offered" => insta::sorted_redaction(),

    });

    // make real metadata request
    assert_let!(Effect::Http(request), &mut update.effects[0]);
    let op = &request.operation;
    assert_eq!(op.url, format!("{}/.well-known/openid-credential-issuer", &offer.credential_issuer));

    let resp = issuer.get("/.well-known/openid-credential-issuer").expect_success().await;
    let metadata: MetadataResponse = resp.json();
    let response = HttpResponse::ok().body(resp.into_bytes()).build();
    let update = app.resolve(request, response).expect("update");

    // check the app emitted an (internal) event to update the model
    let response = ResponseBuilder::ok().body(metadata.clone()).build();
    assert_eq!(update.events, vec![Event::Metadata(Ok(response.clone()))]);

    // ------------------------------------------------
    // Event::Metadata
    //   1. saves metadata to model
    // ------------------------------------------------
    app.update(Event::Metadata(Ok(response)), &mut model);
    assert_snapshot!("metadata", app.view(&model), {
        ".offered" => insta::sorted_redaction(),
        ".offered.*.credential_definition.credentialSubject" => insta::sorted_redaction()
    });

    // ------------------------------------------------
    // Event::Accept
    //   1. emits Event::GetToken
    // ------------------------------------------------
    app.update(Event::Accept, &mut model);
    assert_snapshot!("accept", app.view(&model), {
        ".credentials" => insta::sorted_redaction(),
        ".offered.*.credential_definition.credentialSubject" => insta::sorted_redaction()
    });

    // ------------------------------------------------
    // Event::Pin
    //   1. emits Event::GetToken
    // ------------------------------------------------
    let user_pin = offer_resp.user_pin.expect("user pin should be set");
    let mut update = app.update(Event::Pin(user_pin.clone()), &mut model);
    assert_snapshot!("pin", app.view(&model), {
        ".credentials" => insta::sorted_redaction(),
        ".offered.*.credential_definition.credentialSubject"=> insta::sorted_redaction()
    });

    // ------------------------------------------------
    // Event::GetToken
    //   1. creates a TokenRequest
    //   2. emits Event::Token
    // ------------------------------------------------
    assert_let!(Effect::Http(request), &mut update.effects[0]);
    assert_eq!(request.operation.url, format!("{}/token", &offer.credential_issuer));

    // make real token request
    let body = Body::from_bytes(request.operation.body.clone());
    let form: TokenRequest = body.into_form().await.expect("should deserialize");
    let resp = issuer.post("/token").form(&form).expect_success().await;

    // resolve the app request with a ~~simulated~~ response from the server
    let token: TokenResponse = resp.json();
    let response = HttpResponse::ok().body(resp.into_bytes()).build();
    let update = app.resolve(request, response).expect("update");

    // check the app emitted an (internal) event to update the model
    let response = ResponseBuilder::ok().body(token.clone()).build();
    assert_eq!(update.events, vec![Event::Token(Ok(response.clone()))]);

    // ------------------------------------------------
    // Event::Token
    //   1. saves TokenResponse to model
    //   2. requests verification method from shell
    //   2. emits Event::Proof with verification result
    // ------------------------------------------------
    let mut update = app.update(Event::Token(Ok(response)), &mut model);
    assert_snapshot!("token", app.view(&model), {
        ".credentials" => insta::sorted_redaction(),
        ".offered.*.credential_definition.credentialSubject"=> insta::sorted_redaction()
    });

    assert_let!(Effect::Signer(request), &mut update.effects[0]);
    assert_eq!(request.operation, SignerRequest::Verification);

    // simulate the SignerResponse
    let response = SignerResponse::Verification {
        alg: wallet::alg(),
        kid: wallet::kid(),
    };
    let update = app.resolve(request, response).expect("an update");

    // verify the app emitted a Proof event
    assert_let!(Event::Proof(Ok((alg, kid))), &update.events[0]);

    // ------------------------------------------------
    // Event::Proof
    //   1. saves SignerResult to model
    //   requests JWT signing from shell
    //   2. emits Event::Signed
    // ------------------------------------------------
    let mut update = app.update(Event::Proof(Ok((alg.clone(), kid.clone()))), &mut model);

    assert_snapshot!("proof", app.view(&model), {
        ".credentials" => insta::sorted_redaction(),
        ".offered.*.credential_definition.credentialSubject"=> insta::sorted_redaction()
    });

    // check that the app emitted a Signer request
    assert_eq!(update.events, vec![]);
    assert_let!(Effect::Signer(request), &mut update.effects[0]);
    assert_let!(SignerRequest::Sign(msg), &request.operation);

    // simulate the SignerResponse
    let signature = wallet::sign(msg);
    let response = SignerResponse::Signature(signature);
    let update = app.resolve(request, response).expect("an update");

    // verify the app emitted a Signed event
    assert_let!(Event::Signed(Ok(signed)), &update.events[0]);

    // ------------------------------------------------
    // Event::Signed
    //   1. saves TokenResponse to model
    //   2. emits Event::GetCredential for each credential in offer
    // ------------------------------------------------
    let mut update = app.update(Event::Signed(Ok(signed.to_string())), &mut model);
    assert_snapshot!("signed", app.view(&model), {
        ".credentials" => insta::sorted_redaction(),
        ".offered.*.credential_definition.credentialSubject"=> insta::sorted_redaction()
    });

    assert_let!(Effect::Http(request), &mut update.effects[0]);
    let op = &request.operation;
    assert_eq!(op.url, format!("{}/credential", &offer.credential_issuer));

    // make real credential request
    let body: Value = serde_json::from_slice(&op.body).expect("should deserialize");
    let auth_header = HeaderValue::from_str(&format!("Bearer {}", &token.access_token))
        .expect("should create header");
    let resp = issuer
        .post("/credential")
        .add_header(AUTHORIZATION, auth_header)
        .json(&body)
        .expect_success()
        .await;

    // resolve the app request with a ~~simulated~~ response from the server
    let cred: CredentialResponse = resp.json();
    let response = HttpResponse::ok().body(resp.into_bytes()).build();
    let update = app.resolve(request, response).expect("update");

    // check the app emitted an (internal) event to interpret the response
    let response = ResponseBuilder::ok().body(cred.clone()).build();
    assert_eq!(update.events, vec![Event::Credential(Ok(response.clone()))]);

    // ------------------------------------------------
    // Event::Credential
    //   1. Unpacks Credential from CredentialResponse
    //   2. emits Event::GetLogo
    // ------------------------------------------------

    let mut update = app.update(Event::Credential(Ok(response.clone())), &mut model);
    assert_let!(Effect::Http(_request), &mut update.effects[0]);
    let response =
        ResponseBuilder::ok().body(b"sample".to_vec()).header("Content-Type", "image/png").build();
    let credential = credential::Credential::sample();
    let mut update = app.update(Event::Logo(credential, Ok(response.clone())), &mut model);

    // ------------------------------------------------
    // Event::AddCredential
    //   1. saves Credential to store
    //   2. emits Event::CredentialAdded
    // ------------------------------------------------

    // resolve store.add() output
    assert_let!(Effect::Store(request), &mut update.effects[0]);
    let update = app.resolve(request, StoreResponse::Ok).expect("update");

    // ------------------------------------------------
    // Event::CredentialAdded
    //   1. model is reset
    // ------------------------------------------------
    app.update(update.events[0].clone(), &mut model);
    assert_snapshot!("credential-saved", app.view(&model));
}
