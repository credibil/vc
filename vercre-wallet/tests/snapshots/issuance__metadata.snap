---
source: vercre-wallet/tests/issuance.rs
assertion_line: 83
expression: app.view(&model)
---
issuer: "http://credibil.io"
offered:
  EmployeeID_JWT:
    format: jwt_vc_json
    scope: EmployeeIDCredential
    cryptographic_binding_methods_supported:
      - "did:jwk"
      - "did:ion"
    credential_signing_alg_values_supported:
      - ES256K
      - EdDSA
    proof_types_supported:
      jwt:
        proof_signing_alg_values_supported:
          - ES256K
          - EdDSA
    display:
      - name: Employee ID
        locale: en-NZ
        logo:
          uri: "https://credibil.github.io/assets/credibil-logo-reversed.png"
          alt_text: Credibil Logo
        description: Credibil employee ID credential
        background_color: "#323ed2"
        background_image:
          uri: "https://credibil.github.io/assets/credibil-background.png"
          alt_text: Credibil Background
        text_color: "#ffffff"
    credential_definition:
      "@context":
        - "https://www.w3.org/2018/credentials/v1"
        - "https://www.w3.org/2018/credentials/examples/v1"
      type:
        - VerifiableCredential
        - EmployeeIDCredential
      credentialSubject:
        email:
          mandatory: true
          value_type: string
          display:
            - name: Email
              locale: en-NZ
        familyName:
          mandatory: true
          value_type: string
          display:
            - name: Family name
              locale: en-NZ
        givenName:
          mandatory: true
          value_type: string
          display:
            - name: Given name
              locale: en-NZ
status: Ready
